### ISSUE TITLE
different sample num of nq dev dataset

### ISSUE THREAD

**Issue by sysu-zjw** [facebookresearch/DPR#157]
Hi, I found 6515 samples from `biencoder-nq-dev.json` but 8757 samples from `nq-dev.qa.csv`. Wonder why the number of samples varies? Thanks.

I noticed the model has better performance on these 6515 samples from `biencoder-nq-dev.json` against overall 8757 dev samples. Just curious about the sampling strategies for `biencoder-nq-dev.json`, and would like to confirm if those 2242 samples are discarded only because their positive documents fail to match the different version of Wikipedia (according to session 4.2 of DPR's paper)?

**Comment by vlad-karpukhin** [comment]
Hi,
that is because of gold passage mapping, it is explained in the paper and is also mentioned in one  of the closed issue here.
See https://github.com/facebookresearch/DPR/issues/137 for references.

**Comment by sysu-zjw** [comment]
Thanks, @vlad-karpukhin . I have two additional information want to confirm with you:

1. In #114 , you mention 

> single-adv-hn.nq.bert-base-encoder - is a checkpoint trained by concatenating the original NQ train data set we provided & this new dataset.

What does "concatenating" mean here? Does it mean the training set for the current best performance is nq-train & nq-adv-train (i.e. 58880 + 69639 samples)?


2. If I understand correctly, it still uses only one hard negative for each query for the new adv-DPR training setting. The only difference is that the hard negative is either comes from BM25 or the original DPR result? Besides the different source of hard negative, everything holds the same as the original DPR,  is this correct?

**Comment by vlad-karpukhin** [comment]
Hi,
1. Yes that effectively means you have one datasets set as a union of the tow datasets with identical questions but different passages.
2. I used volta 8x32GB to train that
3. One hard negative is used for every model forward pass per question but each time it is selected randomly from a pool of hard negatives. You can also enable to use distantly supervised positives by setting shuffle_positives: True in conf/datasets/encoder_train_default.yaml configuration for each dataset used.
4. yes, see my answer for 3 above. I don't expect results will change a lot if you enable to use distantly supervised positives.

**Comment by sysu-zjw** [comment]
Thanks, @vlad-karpukhin . It's very clear.
